Fix segmentation fault due to access to destroyed CudaIPCGlobalEntities instance

Flamefire opened the pull request:
There is an instance of the static destruction order fiasco where cuda_ipc_global_entities may be accessed after it is destroyed. See #51961

This change uses a flag and avoids accesses to the destroyed class when it is set to false.

Fixes #51961

This removes the function to clear shared_blocks introduced by #53080 which had multiple issues: Unprotected access to a shared structure and modification of the vector which is being cleared by the destructors of the objects contained.
I.e. what happened was:

- `CudaIPCSentDataLimbo_.clear_shared_blocks();` is called from the destructor of CudaIPCGlobalEntities as of your PR
- This deletes instances of `CudaIPCSentData` which hold `at::DataPtr` created by `GetNewRefCountedSentData`
- This means `CudaIPCSentDataDelete` is called with still active pointers
- Hence `CudaIPCSentDataLimbo_.add` is called adding a new value to `shared_blocks_`

Flamefire committed:
Fix segmentation fault due to access to destroyed CudaIPCGlobalEntities instance

There is an instance of the static destruction order fiasco where cuda_ipc_global_entities
may be accessed after it is destroyed. See #51961

This change uses a flag and avoids accesses to the destroyed class when it is set to false.

VitalyFedyunin commented on torch/csrc/CudaIPCTypes.cpp:
```diff
@@ -25,13 +25,20 @@ void warnProducerTerminatedBeforeSharedTensorsReleased() {
   std::shared_ptr<CudaIPCRefCountersFile> next_available_ref_counters_file_;
   CudaIPCSentDataLimbo CudaIPCSentDataLimbo_;
-  CudaIPCGlobalEntities() : ref_counters_files_() {}
+  CudaIPCGlobalEntities() { alive = true; }
```
nit: maybe leave ref_counters_files_() call?

Flamefire replied:
I did remove it for consistency while I'm already touching that line. `ref_counters_files_` is a map, which has a default ctor, same as mutex and shared_ptr, so if you call its default ctor explicitly then I'd guess the other should be as well, but it doesn't change anything as C++ default constructs those already, so IMO less code is better

Flamefire committed:
Remove clear_shared_blocks call in dtor

Flamefire committed:
Fix race conditions and possible data corruption

Protect all accesses to sharede_blocks via mutex
On clear move the shared_blocks_ first to a temporary buffer as deleting
the elements may add new ones

Flamefire commented:
@pritamdamania87 Rebased to master, resolved conflicts and fixed new issues and the race conditions I've found

VitalyFedyunin commented:
Could you please rebase it and modify with #51961 comments in mind. Thanks.

Flamefire committed:
Remove unused clear_shared_blocks

Flamefire commented:
@VitalyFedyunin Rebased and resolved conflicts
> Could you please rebase it and modify with #51961 comments in mind. Thanks.

Not sure if this is what you meant. While it might be possible to use the `shared_blocks_.clear()` approach there still may be instances slipping through. Also avoiding the newly introduced failures by that (see updated description and the mentioned issue) is hard and especially requires multiple rounds of clearing that vector until it stays empty. So simply letting it get destructed and avoiding any access to it at all (i.e. especially the adding of new entries) is a lot easier.

VitalyFedyunin commented:
> letting it get destructed and avoiding any access to it at all (i.e. especially the adding of new entries) is a lot easier.

I'm fine with this approach. Feel free to update PR so it can land.

yinghai commented:
@Flamefire thanks for working on this. Could you rebase and revert the third_party part, which seems not relevant to the changes?

Flamefire commented:
@yinghai Ouch, sure. Done
